# 2944 Miltiades Vasiliades
Question 3

In [280]:
import sklearn.metrics
import numpy as np 
import copy
from scipy.sparse import csr_matrix

In [281]:
ratings=np.array([
[5,3,0,1,0,0],
[4,2,1,0,0,1],
[3,0,0,1,3,3],
[2,5,1,5,3,4]])
#print(csr_matrix(ratings))

In [282]:
sklearn.metrics.pairwise.cosine_similarity(csr_matrix(ratings),Y=None)

array([[1.        , 0.93697496, 0.51110125, 0.56694671],
       [0.93697496, 1.        , 0.60436722, 0.54824099],
       [0.51110125, 0.60436722, 1.        , 0.6761234 ],
       [0.56694671, 0.54824099, 0.6761234 , 1.        ]])


Observe that user X is similar to Y and W

In [283]:
sixRating1 = np.average([1,4],weights=[0.93697496,0.56694671])
print(sixRating1)
ratingsAmmended=copy.deepcopy(ratings)
ratingsAmmended[0,5] = sixRating1
print(ratingsAmmended)

2.130936646454466
[[5 3 0 1 0 2]
 [4 2 1 0 0 1]
 [3 0 0 1 3 3]
 [2 5 1 5 3 4]]


In [284]:
#columnMeans=np.mean(ratings,axis=0)
#print(columnMeans)
dimensions=ratings.shape
meanRatings=[]
normalisedRatings=copy.deepcopy(ratings)
for User in range(dimensions[0]):
    sm = 0
    nonZero=0
    mean=0
    for Movie in range(dimensions[1]):
        if ratings[User][Movie]!=0:
            nonZero+=1
            sm=sm+ratings[User][Movie]
    mean=sm/nonZero
    meanRatings.append(mean)
    for Movie in range(dimensions[1]):
        if ratings[User][Movie]!=0:
            normalisedRatings[User][Movie]=ratings[User][Movie]-mean
    
print(normalisedRatings)
print(meanRatings)


[[ 2  0  0 -2  0  0]
 [ 2  0 -1  0  0 -1]
 [ 0  0  0 -1  0  0]
 [-1  1 -2  1  0  0]]
[3.0, 2.0, 2.5, 3.3333333333333335]


In [285]:
cosSIMarray2 = sklearn.metrics.pairwise.cosine_similarity(csr_matrix(normalisedRatings),Y=None)
print(cosSIMarray2)

[[ 1.          0.57735027  0.70710678 -0.53452248]
 [ 0.57735027  1.          0.          0.        ]
 [ 0.70710678  0.          1.         -0.37796447]
 [-0.53452248  0.         -0.37796447  1.        ]]


Παρατηρώ οτι  αυτή την φορά ομοιότεροι χρήστες με τον Χ είναι ο Z και o Y με τον Z

In [286]:
Zeta = cosSIMarray2[0][1] + cosSIMarray2[0][2]
print("Zeta="+str(Zeta))
MeanDeviationOfSimilarUsers = ((cosSIMarray2[0][1]*(ratings[1][5]-meanRatings[1])) + (cosSIMarray2[0][2]*(ratings[2][5]-meanRatings[2])))/Zeta
print(MeanDeviationOfSimilarUsers)
RuserX6 = meanRatings[0]+MeanDeviationOfSimilarUsers
DeviationFromMeanUserXMovie6=RuserX6-meanRatings[0]
print(RuserX6)

Zeta=1.2844570503761732
-0.17423461417476724
2.8257653858252327


In [287]:
print('First Method Result for Rating of User X for movie 0 is '+str(sixRating1))
print('Using the Second method the rating is now '+str(RuserX6)+" Thus we calculate the deviation to be"+ str(DeviationFromMeanUserXMovie6))


First Method Result for Rating of User X for movie 0 is 2.130936646454466
Using the Second method the rating is now 2.8257653858252327 Thus we calculate the deviation to be-0.17423461417476727


## General Observations
Using the first method we come to the conclusion than User X is most similar to Users Y and W computing the unknown rating as the weighted average of the rating of these users gave
us a rating of 2.130936646454466 for that movie which we can round it down to 2
Using the second method we come to the conclusion that this time the most similar users to X are Y and Z computing the unknown rating now results in an average rating of 2.8257653858252327 
which can be rounded up to 3!
Therefore it can be concluded that it's not always obvious which users are more similar without first normalising the data
